### PR TITLE
update libsync 

### PR DESCRIPTION
**Pull request by JimmyShi22**
(1) add more unitest
(2) update downloading block queue
(3) add sync demo
(4) modify some bugs

### CODE REVIEW

File: libsync/SyncMaster.cpp [hunk 1]
```diff
@@ -244,17 +255,29 @@ void SyncMaster::maintainPeersStatus()
...
+        // TODO check minerlist (&& minerlist sig is ok)
         {
-            // TODO: verify transaction signature before blockVerifier
-            ExecutiveContext::Ptr exeCtx = m_blockVerifier->executeBlock(*block);
-            m_blockChain->commitBlock(*block, exeCtx);
+            ExecutiveContext::Ptr exeCtx = m_blockVerifier->executeBlock(*topBlock);
+            m_blockChain->commitBlock(*topBlock, exeCtx);
         }
+        else
+            LOG(WARNING) << "Ignore illigal block " << topBlock->header().number();
```
Comment (wheatli) [inline libsync/SyncMaster.cpp]: discard illegal block ?

File: libsync/DownloadingBlockQueue.cpp [hunk 2]
```diff
@@ -61,8 +61,37 @@ size_t DownloadingBlockQueue::size()
...
+    ReadGuard l(x_blocks);
+    if (!m_blocks.empty())
+        return m_blocks.top();
+    else
+        return nullptr;
+}
+
+void DownloadingBlockQueue::clear()
+{
+    ReadGuard l1(x_buffer);
```
Comment (wheatli) [inline libsync/DownloadingBlockQueue.cpp]: It supposed to use WriteGuard

File: libsync/SyncMaster.cpp [hunk 3]
```diff
@@ -202,17 +203,27 @@ void SyncMaster::maintainPeersStatus()
         noteDownloadingBegin();
     }
 
+    // Choose to use min number in blockqueue or max peer number
+    int64_t maxRequestNumber = maxPeerNumber;
+    if (nullptr != m_syncStatus->bq().top())
+    {
+        int64_t minNumberInQueue = m_syncStatus->bq().top()->header().number();
```
Comment (wheatli) [inline libsync/SyncMaster.cpp]: you should save the return ptr of bq().top() but not fetch again

File: libsync/SyncMsgEngine.cpp [hunk 4]
```diff
@@ -130,8 +143,8 @@ void SyncMsgEngine::onPeerBlocks(SyncMsgPacket const& _packet)
     for (unsigned i = 0; i < itemCount; ++i)
     {
         shared_ptr<Block> block = make_shared<Block>(rlps[i].toBytes());
-        // TODO check minerlist sig
-        m_syncStatus->bq().push(block);
+        if (checkImportBlock(block))
```
Comment (wheatli) [inline libsync/SyncMsgEngine.cpp]: it's better to named  isNewBlock

Comment (JimmyShi22) [reply libsync/SyncMsgEngine.cpp]: done

File: libsync/SyncMaster.cpp (hunk 1 above)
Comment (JimmyShi22) [reply libsync/SyncMaster.cpp]: Will correct the mistake word "illigal"
By the way, there may be some illegal blocks in blockqueue, if not isNewBlock(), just pop it, not commit it.

File: libsync/SyncMaster.cpp (hunk 3 above)
Comment (JimmyShi22) [reply libsync/SyncMaster.cpp]: Sure

File: libsync/SyncMaster.cpp (hunk 1 above)
Comment (wheatli) [reply libsync/SyncMaster.cpp]: what kind of situation will lead to illegal blocks pushed into queue?

**Pull request merged**